### PR TITLE
Framework changes for the Script Library Management REST API

### PR DESCRIPTION
**Pull request by Wathsara**
This PR Subjected to - [Script Library REST API](https://github.com/wso2/product-is/issues/7302)

This PR includes :

- Script(Function) library management client and server error handling and management.

- Modified unit tests.

### CODE REVIEW

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/FunctionLibraryManagementAdminService.java [hunk 1]
```diff
@@ -170,10 +174,11 @@ private void evaluateScript(FunctionLibrary functionLibrary) throws FunctionLibr
             code = head + code;
             engine.eval(code);
         } catch (ScriptException e) {
-            log.error("Function library script of " + functionLibrary.getFunctionLibraryName() +
+            log.debug("Script library script of " + functionLibrary.getFunctionLibraryName() +
```
Comment (emswbandara) [inline components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/FunctionLibraryManagementAdminService.java]: need to check if(log.isDebugEnabled)

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/FunctionLibraryManagementServiceImpl.java [hunk 2]
```diff
@@ -168,10 +176,12 @@ private void evaluateScript(FunctionLibrary functionLibrary) throws FunctionLibr
             code = head + code;
             engine.eval(code);
         } catch (ScriptException e) {
-            log.error("Function library script of " + functionLibrary.getFunctionLibraryName() +
+            log.debug("Script library script of " + functionLibrary.getFunctionLibraryName() +
```
Comment (emswbandara) [inline components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/FunctionLibraryManagementServiceImpl.java]: check if debug is enabled

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/dao/impl/FunctionLibraryDAOImpl.java [hunk 3]
```diff
@@ -56,7 +59,7 @@
      * @throws FunctionLibraryManagementException
      */
     public void createFunctionLibrary(FunctionLibrary functionLibrary, String tenantDomain)
-            throws FunctionLibraryManagementException {
+            throws FunctionLibraryManagementServerException {
```
Comment (emswbandara) [inline components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/dao/impl/FunctionLibraryDAOImpl.java]: why this change is required? isn't FunctionLibraryManagementServerException a subclass of FunctionLibraryManagementException?

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/dao/impl/FunctionLibraryDAOImpl.java [hunk 4]
```diff
@@ -105,10 +112,10 @@ public void createFunctionLibrary(FunctionLibrary functionLibrary, String tenant
      * @param functionLibraryName Function library name
      * @param tenantDomain        Tenant domain
      * @return Function library
-     * @throws FunctionLibraryManagementException
+     * @throws FunctionLibraryManagementServerException
      */
     public FunctionLibrary getFunctionLibrary(String functionLibraryName, String tenantDomain)
-            throws FunctionLibraryManagementException {
+            throws FunctionLibraryManagementServerException {
```
Comment (emswbandara) [inline components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/dao/impl/FunctionLibraryDAOImpl.java]: why this change is required? isn't FunctionLibraryManagementServerException a subclass of FunctionLibraryManagementException? In that case, we can keep the method interface as it is

Comment (emswbandara) [reply components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/dao/impl/FunctionLibraryDAOImpl.java]: please check other places as well

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/util/FunctionLibraryExceptionManagementUtil.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    /**
+     * This method can be used to generate a FunctionLibraryManagementClientException from
+     * FunctionLibraryManagementConstants.ErrorMessage object when no exception is thrown.
+     *
+     * @param error FunctionLibraryManagementConstants.ErrorMessage.
+     * @param data  data to replace if message needs to be replaced.
+     * @return FunctionLibraryManagementClientException.
+     */
+    public static FunctionLibraryManagementClientException handleClientException(
+            FunctionLibraryManagementConstants.ErrorMessage
```
Comment (emswbandara) [inline components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/util/FunctionLibraryExceptionManagementUtil.java]: shall we inline line 38 and 39

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/FunctionLibraryManagementAdminService.java (hunk 1 above)
Comment (Wathsara) [reply components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/FunctionLibraryManagementAdminService.java]: My bad. It should be a log.error() not log.debug.

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/FunctionLibraryManagementServiceImpl.java (hunk 2 above)
Comment (Wathsara) [reply components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/FunctionLibraryManagementServiceImpl.java]: My bad. It should be a log.error() not log.debug.

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/FunctionLibraryManagementAdminService.java (hunk 1 above)
Comment (Wathsara) [reply components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/FunctionLibraryManagementAdminService.java]: [Fixed Commit ](https://github.com/wso2/carbon-identity-framework/pull/2679/commits/9a526929402c0b6fb5ef8eab5d87a89fbaeee7c8)

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/util/FunctionLibraryExceptionManagementUtil.java (hunk 5 above)
Comment (Wathsara) [reply components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/util/FunctionLibraryExceptionManagementUtil.java]: [Fixed Commit](https://github.com/wso2/carbon-identity-framework/pull/2679/commits/9a526929402c0b6fb5ef8eab5d87a89fbaeee7c8)

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/dao/impl/FunctionLibraryDAOImpl.java (hunk 4 above)
Comment (Wathsara) [reply components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/dao/impl/FunctionLibraryDAOImpl.java]: [Fixed commit](https://github.com/wso2/carbon-identity-framework/pull/2679/commits/9a526929402c0b6fb5ef8eab5d87a89fbaeee7c8)

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/dao/impl/FunctionLibraryDAOImpl.java (hunk 3 above)
Comment (Wathsara) [reply components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/dao/impl/FunctionLibraryDAOImpl.java]: [Fixed commit](https://github.com/wso2/carbon-identity-framework/pull/2679/commits/9a526929402c0b6fb5ef8eab5d87a89fbaeee7c8)

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/FunctionLibraryManagementServiceImpl.java (hunk 2 above)
Comment (Wathsara) [reply components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/FunctionLibraryManagementServiceImpl.java]: [Fixed commit](https://github.com/wso2/carbon-identity-framework/pull/2679/commits/9a526929402c0b6fb5ef8eab5d87a89fbaeee7c8)

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/util/FunctionLibraryExceptionManagementUtil.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+     *
+     * @param error FunctionLibraryManagementConstants.ErrorMessage.
+     * @param data  data to replace if message needs to be replaced.
+     * @return message format with data.
+     */
+    private static String includeData(FunctionLibraryManagementConstants.ErrorMessage error, String data) {
+
+        String message;
+        if (StringUtils.isNotBlank(data)) {
+            message = String.format(error.getMessage(), data);
```
Comment (pulasthi7) [inline components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/util/FunctionLibraryExceptionManagementUtil.java]: This will be bit consuming and error prone. We assume the error should have exactly one string place holder which we can't enforce. Would be better to give the responsibility to the callee to format/build and pass the exact message.

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/util/FunctionLibraryManagementConstants.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+ * Function Library Management constant class.
+ */
+public class FunctionLibraryManagementConstants {
+
+    /**
+     * Enum for error messages.
+     */
+    public enum ErrorMessage {
+
+        ERROR_CODE_UNEXPECTED("SCL-65001", "Unexpected Error"),
```
Comment (pulasthi7) [inline components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/util/FunctionLibraryManagementConstants.java]: I Guess the code should start with matching HTTP code. @ayshsandu , @IsuraD Can you please comment?

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/util/FunctionLibraryManagementConstants.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        ERROR_CODE_INVALID_SCRIPT_LIBRARY_NAME("SCL-65004", "The script library name is not valid! It is not adhering" +
+                " to the regex %s."),
+        ERROR_CODE_ADD_SCRIPT_LIBRARY("SCL-65002", "Error while creating the script library: %s."),
+        ERROR_CODE_PROCESSING_CONTENT_STREAM_SCRIPT_LIBRARY("SCL-65003",
+                "An error occurred while processing content stream of script library script: %s."),
+        ERROR_CODE_INVALID_TENANT("SCL-65004", "Error while creating script library due to invalid tenant."),
+        ERROR_CODE_RETRIEVE_SCRIPT_LIBRARY("SCL-65005", "Error while getting the script library: %s."),
+        ERROR_CODE_RETRIEVE_SCRIPT_LIBRARIES("SCL-65006", "Error while reading Script libraries."),
+        ERROR_CODE_UPDATE_SCRIPT_LIBRARY("SCL-65007", "Failed to update Script library: %s."),
+        ERROR_CODE_DELETE_SCRIPT_LIBRARY("SCL-65008", "Error while deleting Script library: %s."),
```
Comment (pulasthi7) [inline components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/util/FunctionLibraryManagementConstants.java]:
```suggestion
        ERROR_CODE_DELETE_SCRIPT_LIBRARY("SCL-65008", "Error while deleting script library: %s."),
```

Check and fix other similar usages

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/test/java/org/wso2/carbon/identity/functions/library/mgt/FunctionLibraryManagementServiceTest.java [hunk 9]
```diff
@@ -123,28 +123,28 @@ public void createFunctionLibrary(Object functionLibrary, String tenantDomain) {
...
             assertEquals(functionLibraryManagementService.getFunctionLibrary(
                     ((FunctionLibrary) functionLibrary).getFunctionLibraryName(), tenantDomain)
-                                                         .getFunctionLibraryName(),
+                            .getFunctionLibraryName(),
                     ((FunctionLibrary) functionLibrary).getFunctionLibraryName());
 
             // Clean after test
             functionLibraryManagementService.deleteFunctionLibrary(
                     ((FunctionLibrary) functionLibrary).getFunctionLibraryName(), tenantDomain);
         } catch (FunctionLibraryManagementException e) {
```
Comment (pulasthi7) [inline components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/test/java/org/wso2/carbon/identity/functions/library/mgt/FunctionLibraryManagementServiceTest.java]: add a fail() before the catch, if we expect delete call to throw an exception

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/test/java/org/wso2/carbon/identity/functions/library/mgt/FunctionLibraryManagementServiceTest.java [hunk 10]
```diff
@@ -123,28 +123,28 @@ public void createFunctionLibrary(Object functionLibrary, String tenantDomain) {
...
             if (((FunctionLibrary) functionLibrary).getFunctionLibraryName() == null) {
-                assertEquals(e.getMessage(), "Function Library Name is required");
+                assertEquals(e.getMessage(), "script Library Name is required");
             } else if (!isRegexValidated(((FunctionLibrary) functionLibrary).getFunctionLibraryName())) {
-                assertEquals(e.getMessage(), "The function library name " + ((FunctionLibrary) functionLibrary).
-                        getFunctionLibraryName() + " is not valid! It is not adhering " +
-                        "to the regex " + FunctionLibraryMgtUtil.FUNCTION_LIBRARY_NAME_VALIDATING_REGEX + ".");
+                assertEquals(e.getMessage(), "The script library name is not valid! It is not adhering to the regex " +
```
Comment (pulasthi7) [inline components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/test/java/org/wso2/carbon/identity/functions/library/mgt/FunctionLibraryManagementServiceTest.java]: Better to split these to multiple tests for clarity

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/test/java/org/wso2/carbon/identity/functions/library/mgt/FunctionLibraryManagementServiceTest.java (hunk 9 above)
Comment (Wathsara) [reply components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/test/java/org/wso2/carbon/identity/functions/library/mgt/FunctionLibraryManagementServiceTest.java]: [Fixed commit](https://github.com/wso2/carbon-identity-framework/pull/2679/commits/8c5deb0813742b0687efcb1daedf5123e568cb42)

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/util/FunctionLibraryManagementConstants.java (hunk 8 above)
Comment (Wathsara) [reply components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/util/FunctionLibraryManagementConstants.java]: [Fixed commit](https://github.com/wso2/carbon-identity-framework/pull/2679/commits/8c5deb0813742b0687efcb1daedf5123e568cb42)

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/util/FunctionLibraryManagementConstants.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+     * Enum for error messages.
+     */
+    public enum ErrorMessage {
+
+        ERROR_CODE_UNEXPECTED("SCL-65001", "Unexpected Error"),
+        ERROR_CODE_DATABASE_CONNECTION("SCL-65001", "Couldn't get a database connection."),
+        ERROR_CODE_VALIDATE_SCRIPT_LIBRARY_SCRIPT("SCL-60002", "Script library script of %s contains errors."),
+        ERROR_CODE_REQUIRE_SCRIPT_LIBRARY_NAME("SCL-60003", "Script library name is required"),
+        ERROR_CODE_REQUIRE_SCRIPT_LIBRARY_SCRIPT("SCL-60003", "Script library script is required"),
+        ERROR_CODE_ALL_READY_EXIST_SCRIPT_LIBRARY("SCL-65004", "Already a script library available with the name: %s."),
```
Comment (emswbandara) [inline components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/util/FunctionLibraryManagementConstants.java]: Typo.. change ALL_READY to ALREADY

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/util/FunctionLibraryManagementConstants.java (hunk 7 above)
Comment (emswbandara) [reply components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/util/FunctionLibraryManagementConstants.java]: This format was established for server and client errors

File: components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/util/FunctionLibraryManagementConstants.java (hunk 11 above)
Comment (Wathsara) [reply components/functions-library-mgt/org.wso2.carbon.identity.functions.library.mgt/src/main/java/org/wso2/carbon/identity/functions/library/mgt/util/FunctionLibraryManagementConstants.java]: [Fixed commit](https://github.com/wso2/carbon-identity-framework/pull/2679/commits/3d6fa4d553b0d78ca83f8d85f66d01572a985143)
